core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -621,55 +715,108 @@ public boolean deleteFromAllStores(Object key, int segment, Predicate<? super St
          return Flowable.using(publisherSemaphoreCallable, semaphore -> {
             semaphore.acquire();
             return segmentedStore.publishKeys(segments, filter);
-         }, Semaphore::release);
+         }, Semaphore::release)
+               .subscribeOn(persistenceScheduler);
       }
 
       return publishKeys(PersistenceUtil.combinePredicate(segments, keyPartitioner, filter), predicate);
    }
 
    @Override
-   public MarshallableEntry loadFromAllStores(Object key, boolean localInvocation, boolean includeStores) {
+   public <K, V> MarshallableEntry<K, V> loadFromAllStoresSync(Object key, boolean localInvocation, boolean includeStores) {
+      // Some code calls into PersistenceManager even if there are no stores configured
+      if (loaders.isEmpty()) {
+         return null;
+      }
+      assert Thread.currentThread().getName().startsWith("persistence") : "Thread name is: " + Thread.currentThread().getName();
+      return loadFromAllStoresSync(key, localInvocation, includeStores, -1);
+   }
+
+   private <K, V> MarshallableEntry<K, V> loadFromAllStoresSync(Object key, boolean localInvocation, boolean includeStores, int traceId) {
       storesMutex.readLock().lock();
       try {
          checkStoreAvailability();
+         if (trace) {
+            log.tracef("Loading entry for key %s from stores with includeStores %s for id: %d",
+                  key, includeStores, traceId);
+         }
+         MarshallableEntry load = null;
          for (CacheLoader l : loaders) {
             if (allowLoad(l, localInvocation, includeStores)) {
-               MarshallableEntry load = l.loadEntry(key);
+               load = l.loadEntry(key);
                if (load != null)
-                  return load;
+                  break;
             }
          }
-         return null;
+         if (trace) {
+            log.tracef("Entry was loaded: %s for key %s from stores for id: %d", load, key, traceId);
+         }
+         return load;
       } finally {
          storesMutex.readLock().unlock();
       }
    }
 
    @Override
-   public MarshallableEntry loadFromAllStores(Object key, int segment, boolean localInvocation, boolean includeStores) {
+   public CompletionStage<MarshallableEntry> loadFromAllStores(Object key, boolean localInvocation, boolean includeStores) {
```

Comment:
`<K, V> CompletionStage<MarshallableEntry<K,V>>`?

Reply:
Yeah, I can't remember why I didn't put the generics on here, cause I remember specifically not doing it for a method. I will look at it again.